File: kotlin/nitroping/src/main/kotlin/dev/nitroping/Events.kt
/*
 * Events.kt
 *
 * `np.events` — engagement reporting.
 *
 * Posted from the client app when the user opens or taps an action on a
 * notification, so the dashboard can show open / click counters next to
 * delivered counters.
 *
 * Wire: `POST /api/v1/events` with body
 *     { "notification_id": "...", "type": "opened" | "clicked",
 *       "device_id": "...", "action_id": "..." }
 *
 * The server returns a 202 with `{"id":"evt_...","accepted":true}`.
 */

package dev.nitroping

import dev.nitroping.internal.HttpTransport

/** The two kinds of engagement we report back to nitroping. */
public enum class EngagementType(public val wire: String) {
    /** The user tapped the notification body (or the default action). */
    OPENED("opened"),

    /** The user tapped one of the action buttons; pair with `actionId`. */
    CLICKED("clicked"),
}

public class Events internal constructor(private val transport: HttpTransport) {
    /**
     * Report an engagement event.
     *
     * @param notificationId server-side notification id (`notif_...`).
     * @param type [EngagementType.OPENED] or [EngagementType.CLICKED].
     * @param deviceId server-side device id (`dev_...`). Optional but
     *                 recommended; lets the dashboard slice by device.
     * @param actionId action button id when [type] is [EngagementType.CLICKED].
     */
    public suspend fun report(
        notificationId: String,
        type: EngagementType,
        deviceId: String? = null,
        actionId: String? = null,
    ) {
        require(notificationId.isNotEmpty()) { "notificationId must not be empty" }
        val body = LinkedHashMap<String, Any?>()
        body["notification_id"] = notificationId
        body["type"] = type.wire
        if (deviceId != null) body["device_id"] = deviceId
        if (actionId != null) body["action_id"] = actionId
        transport.request(method = "POST", path = "/api/v1/events", body = body)
    }
}

File: python/src/nitroping/_events.py
"""``events`` resource client.

Mounted on :class:`nitroping.Nitroping` as ``np.events``. Wraps
``POST /api/v1/events`` — the public, unauthenticated engagement
endpoint. The ``(notification_id, device_id)`` pair is the bearer
secret, so no ``Authorization`` header is required (and a ``pk_`` public
key is fine).

This is the endpoint a client app calls when a notification is opened or
a notification action is clicked.
"""

from __future__ import annotations

from typing import Any, cast

from ._http import HttpClient
from .types import EngagementEvent, ReportEventResult


class EventsClient:
    """Report public engagement events (``opened`` / ``clicked``)."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def report(
        self,
        *,
        notification_id: str,
        device_id: str,
        type: EngagementEvent,
        action_id: str | None = None,
        happened_at: str | None = None,
    ) -> ReportEventResult:
        """Report an engagement event (``opened`` or ``clicked``).

        Returns ``{"accepted": True}`` on 202. Raises
        :class:`~nitroping.errors.ApiError` with ``code = "not_found"``
        (404) if the notification/device pair is unknown.
        """
        wire: dict[str, Any] = {
            "notification_id": notification_id,
            "device_id": device_id,
            "type": type,
        }
        if action_id is not None:
            wire["action_id"] = action_id
        if happened_at is not None:
            wire["happened_at"] = happened_at

        response = self._http.request("POST", "/api/v1/events", body=wire)
        return cast(ReportEventResult, response)
